Enforce connection preconditions for setParent

jschanker opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves #4989 

<!-- TODO: What Github issue does this resolve? Please include a link. -->

### Proposed Changes

<!-- TODO: Describe what this Pull Request does.  Include screenshots if applicable. -->
Throws error when attempting to set a block's parent to a non-null block `b` if `b` is `not` properly connected to it without changing state (i.e., if its previous/output connection is *not* connected to one of `b`'s input/next connections or when it doesn't have such connections).  Also throws error when attempting to set a block's parent to `null` if it has a connection to a superior block.

#### Behavior Before Change

Setting a parent of a block already connected to some other superior block would remove the block from that old parent's child list before throwing, leading to potential rendering issues.  Setting the parent of a disconnected block could cause rendering issues and errors when attempting to delete that parent.

<!--TODO: Image, gif or explanation of behavior before this pull request. -->

#### Behavior After Change

Errors are thrown in these cases without changing state so none of these problems ensue.

<!--TODO: Image, gif or explanation of behavior after this pull request. -->

### Reason for Changes

<!--TODO: Explain why these changes should be made.  Include screenshots if applicable. -->

To enforce the desired precondition for `setParent` that a non-`null` inputted parent should already be a superior block via the proper connections and that a block should not be connected to a superior one when attempting to set its new parent to `null`.

### Test Coverage

Tested on Blockly Playground to make sure that it resolved the desired bugs.

### Documentation

No additional documentation required.

BeksOmega commented:
Hi jschanker, thanks for posting this! I'm just about to head out for today, so I'll get a review back to you on Monday =)

Also sorry about creating all of those linked issues when I was testing some automations stuff (whoops) I'll see if I can get those cleaned up.

BeksOmega commented on core/block.js:
```diff
@@ -741,21 +741,31 @@ Blockly.Block.prototype.getChildren = function(ordered) {
  */
 Blockly.Block.prototype.setParent = function(newParent) {
-  if (newParent == this.parentBlock_) {
+  if (newParent === this.parentBlock_) {
```
Thanks for fixing this while you were in the area :D

jschanker replied:
Happy to do so!

BeksOmega commented on core/block.js:
```diff
@@ -741,21 +741,31 @@ Blockly.Block.prototype.getChildren = function(ordered) {
+        ' superior block.');
+  }
+
   if (this.parentBlock_) {
```
I know this isn't exactly related to what you're doing, but do you think this if-statement and the below one could be combined? It seems like they're doing the same things (updating child blocks and top blocks).

jschanker replied:
I could be missing something, but I'm not sure how to cleanly group those two if statements given the overlap of the following conditional tasks.

```
                             Remove from old parent  Add to new parent Remove top block  Add top block
Has old/new parent block              Y                     Y                 N               N
Only has old parent block             Y                     N                 N               Y
Only has new parent block             N                     Y                 Y               N               
```

So for an example, having an if statement selecting only for the first case would mean duplicating code to remove/add parents.  But maybe that's fine?  I could also remove the second else and instead conditionally add a top block if there's no new parent by nesting an if statement inside the first, but I'm not sure that makes the code cleaner?  Thoughts?

I did however remove the unnecessary statement setting `this.parentBlock_` to `null` (and associated comment) as it unconditionally gets set to `newParent` later.

BeksOmega reviewed (changes requested):
Thanks for putting up this fix :D I think it's good to enforce these changes programmatically in addition to fixing the visibility.

I think the only thing this change really needs is some tests. You can either choose to add them, or [file an issue](https://github.com/google/blockly/issues/new?assignees=&labels=type%3A+bug%2C+triage&template=bug_report.md) so that someone can add them later. I think there are 5 cases that need to be covered:
  * Properly setting the parent to null
  * Properly setting the parent to a new not-null parent
  * Cases for each of your error situations

But if you think of others that's rad.

The best place to add these tests is probably in the [block_test.js mocha file](https://github.com/google/blockly/blob/master/tests/mocha/block_test.js). I can't remember if you've written tests before, so here are the [mocha docs](https://mochajs.org/#tdd), and the [chai docs](https://www.chaijs.com/api/assert/) just in case. You will probably want to assert [equal](https://www.chaijs.com/api/assert/#method_equal), [throws](https://www.chaijs.com/api/assert/#method_throws), and maybe [doesNotThrow](https://www.chaijs.com/api/assert/#method_throws).

Thank you again for sending this in! The code looks great :D

jschanker commented:
> Thanks for putting up this fix :D I think it's good to enforce these changes programmatically in addition to fixing the visibility.

Happy to help!

> I think the only thing this change really needs is some tests. 

Absolutely!  Added the tests.

> I can't remember if you've written tests before

I have for https://github.com/google/blockly/pull/4876, and I hope it's okay that I cleaned up the ones I wrote in this pull request (creating the workspace in the setup and tearing it down upon completion).

> Thank you again for sending this in! The code looks great :D

My pleasure again and thanks for reviewing!

BeksOmega commented on tests/mocha/block_test.js:
```diff
@@ -890,6 +890,96 @@ suite('Blocks', function() {
+        chai.assert.throws(this.textBlock.setParent
+            .bind(this.textBlock, this.printBlock));
+        this.textJoinBlock.outputConnection.disconnect();
+        // when new parent has available input connection
```
```suggestion
        // When new parent has available input connection.
```

BeksOmega commented on tests/mocha/block_test.js:
```diff
@@ -890,6 +890,96 @@ suite('Blocks', function() {
+      });
+      test('Setting to new parent when orphan', function() {
+        this.textBlock.outputConnection.disconnect();
+        // when new parent has no available input connection
```
```suggestion
        // When new parent has no available input connection.
```

BeksOmega commented on tests/mocha/block_test.js:
```diff
@@ -890,6 +890,96 @@ suite('Blocks', function() {
+            .bind(this.textJoinBlock, this.printBlock));
+        assertNonParentAndOrphan(this.printBlock, this.textJoinBlock, 'TEXT');
+      });
+      test('Setting to new parent when orphan', function() {
```
I like how you tested both with available inputs and without :D

jschanker replied:
Thanks!

BeksOmega commented on tests/mocha/block_test.js:
```diff
@@ -890,6 +890,96 @@ suite('Blocks', function() {
+        chai.assert.throws(this.textBlock.setParent
+            .bind(this.textBlock, this.printBlock));
+        this.textJoinBlock.outputConnection.disconnect();
+        // setting to block with available input connection
```
```suggestion
        // Setting to block with available input connection.
```

BeksOmega reviewed (commented):
Thanks for adding tests, they look great! Just a few formatting changes, then should be gtg :D I think there should be some UI for you to commit them automagically.

> I hope it's okay that I cleaned up the ones I wrote in this pull request (creating the workspace in the setup and tearing it down upon completion).

Oh yeah definitely! In my opinion it's way better to clean up code than to worry about being sticklery on PRs.

jschanker commented:
Great, made the formatting changes.  Thanks again!

BeksOmega reviewed (approved):
LGTM! Thanks again for adding these checks :D
